### PR TITLE
Add further tests

### PR DESCRIPTION
**Pull request by rhuairahrighairidh**
Contains various tests for the api.
Not all passing.

Also contains fix for api.remove state modification. May have been fixed on another branch.

### CODE REVIEW

File: src/__tests__/add-remove.test.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+})
+test('add with nonsense lnd credentials throws', async t => {
+  await t.throwsAsync(
+    t.context.add({
+      settlerType: SettlementEngineType.Lnd,
+      hostname: 'nonsense',
+      tlsCert: 'nonsense',
+      macaroon: 'nonsense',
+      grpcPort: 12345
+    })
```
Comment (kincaidoneil) [inline src/__tests__/add-remove.test.ts]: Maybe add a test for a correct hostname & port, but either the macaroon or cert are invalid? (one for both). That way it forces it create a connection, which should fail.

File: src/__tests__/add-remove.test.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+// TODO expand input validation tests
+test('add with nonsense xrp secret throws', async t => {
+  await t.throwsAsync(
+    t.context.add({
+      settlerType: SettlementEngineType.XrpPaychan,
+      secret: 'this is not an xrp secret'
+    })
+  )
```
Comment (kincaidoneil) [inline src/__tests__/add-remove.test.ts]: Add test for a valid private key but invalid XRP account/insufficient funds on-chain

File: src/__tests__/disconnect.test.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+})
+
+test('after add eth', async t => {
+  const uplink = await addEth(1)(t.context)
+  await t.notThrowsAsync(t.context.disconnect())
+})
+
+test('after deposit eth', async t => {
+  const uplink = await addEth(1)(t.context)
+  await t.notThrowsAsync(t.context.disconnect())
```
Comment (kincaidoneil) [inline src/__tests__/disconnect.test.ts]: Does it deposit to the uplink?

File: src/__tests__/stream-same-asset.test.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+test.afterEach(async t => t.context.disconnect())
+
+test('xrp -> xrp different credentials', testExchange(addXrp(1), addXrp(2)))
+test('eth -> eth different credentials', testExchange(addEth(1), addEth(2)))
+test('btc -> btc different credentials', testExchange(addBtc(2), addBtc(1)))
+
+test('xrp -> xrp same credentials', testExchange(addXrp(1), addXrp(1)))
+test('eth -> eth same credentials', testExchange(addEth(1), addEth(1)))
+test('btc -> btc same credentials', testExchange(addBtc(1), addBtc(1)))
```
Comment (kincaidoneil) [inline src/__tests__/stream-same-asset.test.ts]: Still not entirely sold on these. I feel like they're pretty redundant with the existing funding/exchange tests (or at least we could reduce all 6 into 3).

Comment (kincaidoneil) [reply src/__tests__/stream-same-asset.test.ts]: Another possibility, building on your idea for two sets of tests (down the road): spin up private testnets that are used for most of the tests for speed, and have separate _true_ integration tests using the live testnet?

**Comment by kincaidoneil** [comment]
(I'll submit a PR to fix the XRP add-remove test -- and require 1 less on-chain tx).

File: src/__tests__/disconnect.test.ts (hunk 3 above)
Comment (rhuairahrighairidh) [reply src/__tests__/disconnect.test.ts]: oops

File: src/__tests__/stream-same-asset.test.ts (hunk 4 above)
Comment (rhuairahrighairidh) [reply src/__tests__/stream-same-asset.test.ts]: I've removed these and added a test to verify duplicate uplinks cannot be added.

File: src/__tests__/add-remove.test.ts (hunk 2 above)
Comment (rhuairahrighairidh) [reply src/__tests__/add-remove.test.ts]: I added a test for a non activated account.

File: src/__tests__/add-remove.test.ts (hunk 1 above)
Comment (rhuairahrighairidh) [reply src/__tests__/add-remove.test.ts]: added a test with invalid tls cert and macaroon

File: src/__tests__/add-remove.test.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  // Note: if the secret is correct length but contains invalid characters, an invalid length error is thrown ('private key length is invalid').
+})
+// Test valid lnd uri, but invalid credentials.
+test('add with invalid lnd credentials throws', async t => {
+  await t.throwsAsync(
+    t.context.add({
+      settlerType: SettlementEngineType.Lnd,
+      hostname: process.env[`LIGHTNING_LND_HOST_CLIENT_1`]!,
```
Comment (kincaidoneil) [inline src/__tests__/add-remove.test.ts]:
```suggestion
      hostname: process.env.LIGHTNING_LND_HOST_CLIENT_1!,
```

File: src/__tests__/add-remove.test.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      grpcPort: parseInt(process.env[`LIGHTNING_LND_GRPCPORT_CLIENT_1`]!, 10),
```
Comment (kincaidoneil) [inline src/__tests__/add-remove.test.ts]:
```suggestion
      grpcPort: parseInt(process.env.LIGHTNING_LND_GRPCPORT_CLIENT_1!, 10),
```

**Pull request merged**